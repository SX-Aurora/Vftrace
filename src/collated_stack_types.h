#ifndef COLLATED_STACK_TYPES_H
#define COLLATED_STACK_TYPES_H

#include <stdint.h>

#include <stdbool.h>

#include "stack_types.h"
#include "collated_profiling_types.h"

typedef struct {
   int ngids;
   int maxgids;
   int *gids;
} gid_list_t;

typedef struct {
   // local stack
   vftr_stack_t *local_stack;
   // profiling data for collective logfile
   collated_profile_t profile;
   // gloabl stackID
   int gid;
   // this list is only needed for grouped stacks
   gid_list_t gid_list;
   bool precise;
   int caller;
   int ncallees;
   int *callees;
   // need a copy of the name for even for local functions
   // because functions from other processes might
   // not appear in the local symbol table (e.g. two binaries)
   char *name;
   // hash
   uint64_t hash;
} collated_stack_t;

typedef struct {
   int nstacks;
   int maxstacks;
   collated_stack_t *stacks;
   // whether this is a name grouped stack or regular
   bool namegrouped;
} collated_stacktree_t;

#endif
