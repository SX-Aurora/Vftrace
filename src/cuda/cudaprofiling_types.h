#ifndef CUDAPROFILING_TYPES_H
#define CUDAPROFILING_TYPES_H

#include <stdint.h>
#include <stdbool.h>
#include <cuda_runtime_api.h>

typedef struct {
   cudaEvent_t start;
   cudaEvent_t stop;
   int cbid;
   int n_calls;
   float t_ms;
   size_t memcpy_bytes[2];
   long long overhead_nsec;
} cudaprofile_t;
#endif
