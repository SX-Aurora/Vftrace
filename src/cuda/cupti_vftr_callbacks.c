#include <cuda_runtime_api.h>
#include <cupti.h>

#include "thread_types.h"
#include "threadstack_types.h"
#include "stack_types.h"
#include "profiling_types.h"
#include "vftrace_state.h"
#include "timer.h"
#include "hashing.h"
#include "self_profile.h"

#include "threads.h"
#include "threadstacks.h"
#include "profiling.h"
#include "callprofiling.h"

#include "cudaprofiling.h"
#include "cupti_vftr_callbacks.h"

void vftr_get_cuda_memory_info (CUpti_CallbackId cbid, const CUpti_CallbackData *cb_info,
                           int *mem_dir, size_t *copied_bytes) {

   if (cbid != CUPTI_RUNTIME_TRACE_CBID_cudaMemcpy_v3020 &&
       cbid != CUPTI_RUNTIME_TRACE_CBID_cudaMemcpyAsync_v3020) {
          *mem_dir = CUDA_NOCOPY;
          *copied_bytes = 0;
   } else {
      enum cudaMemcpyKind kind;
      if (cbid == CUPTI_RUNTIME_TRACE_CBID_cudaMemcpy_v3020) {
         *copied_bytes = (size_t)((cudaMemcpy_v3020_params*)(cb_info->functionParams))->count;
         kind = ((cudaMemcpy_v3020_params*)(cb_info->functionParams))->kind;
      } else {
         *copied_bytes = (size_t)((cudaMemcpyAsync_v3020_params*)(cb_info->functionParams))->count;
         kind = ((cudaMemcpyAsync_v3020_params*)(cb_info->functionParams))->kind;
      }
      *mem_dir = kind == cudaMemcpyHostToDevice ? CUDA_COPY_IN : CUDA_COPY_OUT;
   }
}

void vftr_cuda_region_begin (int cbid, const CUpti_CallbackData *cb_info) {
   SELF_PROFILE_START_FUNCTION;
   long long region_entry_time_begin = vftr_get_runtime_nsec();
   thread_t *my_thread = vftr_get_my_thread(&(vftrace.process.threadtree));
   threadstack_t *my_threadstack = vftr_get_my_threadstack(my_thread);

   // Note that this callback function is called everytime a CUDA (or OpenACC) function
   // is evoked. This means that CUDA functions for non-instrumented functions appear
   // with the same caller function. For example, a main function calling a cuBLAS DGEMM
   // will evoke several CUDA calls, of which none will call the default cyg-hook. In order
   // to not assign all CUDA calls to the same function, we need to identify them uniquely.
   // This cannot be done by the CBID alone, since different kernel launches will have the
   // same CBID. Therefore, for kernel launches, we compute a hash from the function name,
   // and add it to the CBID to create a pseudo-address. This is given to update the thread
   // stack as it would be for a normal address read from a symbol table. 

   const char *func_name;
   uint64_t pseudo_addr;
   if (cbid == CUPTI_RUNTIME_TRACE_CBID_cudaLaunch_v3020 || cbid == CUPTI_RUNTIME_TRACE_CBID_cudaLaunchKernel_v7000) {
         func_name = cb_info->symbolName;
         pseudo_addr = vftr_jenkins_murmur_64_hash (strlen(func_name), (uint8_t*)func_name);
   } else {
         func_name = cb_info->functionName;
         pseudo_addr = 0;
   }
   pseudo_addr += cbid;

   my_threadstack = vftr_update_threadstack_region (my_threadstack, my_thread,
                                                    (uintptr_t)pseudo_addr, func_name,
                                                    &vftrace, false);

   stack_t *my_new_stack = vftrace.process.stacktree.stacks + my_threadstack->stackID;
   profile_t *my_profile = vftr_get_my_profile(my_new_stack, my_thread);
   cudaprofile_t *cudaprof = &my_profile->cudaprof;

   // We fill the callprofile and the cudaprofile with the same time measurements obtained
   // from CUpti. When using the Vftrace timer for the callprofile, in any case the
   // overhead of the CUpti calls is included in some way. This gives a wrong impression
   // of the time spent. That's why below we pass 0 to accumulate_callprofiling, as the
   // exit function will supply the time difference.

   // cb_info contains the same information regarding copied bytes for Memcpy events. Here, we
   // accumulate zero. The memory information is retrieved in the exit hook.
   cudaEventRecord(cudaprof->start, 0);
   vftr_accumulate_cudaprofiling (cudaprof, cbid, 1, 0, CUDA_NOCOPY, 0);
   vftr_accumulate_callprofiling (&(my_profile->callprof), 1, 0);
   
   vftr_accumulate_cudaprofiling_overhead (&(my_profile->cudaprof),
                  vftr_get_runtime_nsec() - region_entry_time_begin);
   SELF_PROFILE_END_FUNCTION;
}

void vftr_cuda_region_end (int cbid, const CUpti_CallbackData *cb_info) {
   SELF_PROFILE_START_FUNCTION;
   (void)cb_info;
   long long region_exit_time_begin = vftr_get_runtime_nsec();

   thread_t *my_thread = vftr_get_my_thread(&(vftrace.process.threadtree));
   threadstack_t *my_threadstack = vftr_get_my_threadstack(my_thread);
   stack_t *my_stack = vftrace.process.stacktree.stacks + my_threadstack->stackID;
   profile_t *my_profile = vftr_get_my_profile(my_stack, my_thread);
   cudaprofile_t *cudaprof = &my_profile->cudaprof;

   cudaEventRecord(cudaprof->stop, 0);
   cudaEventSynchronize(cudaprof->stop);
   float t;
   cudaEventElapsedTime(&t, cudaprof->start, cudaprof->stop);

   int mem_dir;
   size_t copied_bytes;
   vftr_get_cuda_memory_info (cbid, cb_info, &mem_dir, &copied_bytes);

   vftr_accumulate_cudaprofiling(cudaprof, cbid, 0, t, mem_dir, copied_bytes);
   // Convert ms -> ns
   vftr_accumulate_callprofiling(&(my_profile->callprof), 0, (long long)(t * 1000000));
   (void)vftr_threadstack_pop(&(my_thread->stacklist));
   vftr_accumulate_cudaprofiling_overhead (&(my_profile->cudaprof),
                   vftr_get_runtime_nsec() - region_exit_time_begin);
   SELF_PROFILE_END_FUNCTION;
}

// The default CUpti callback handle. It vetoes some CBIDs
// and then just calls the corresponding entry or exit region.
void CUPTIAPI vftr_cupti_event_callback (void *userdata,
                                         CUpti_CallbackDomain domain,
                                         CUpti_CallbackId cbid,
 					 const CUpti_CallbackData *cb_info) {
   // These functions are called in the profiling layer. We need to exclude them,
   // otherwise the callback will be called infinitely
   switch (cbid) {
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventRecord_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventRecordWithFlags_v11010:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventSynchronize_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventElapsedTime_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventCreate_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventCreateWithFlags_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventDestroy_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaEventQuery_v3020:
      case CUPTI_RUNTIME_TRACE_CBID_cudaGetDeviceProperties_v3020:
         return;
   }

   if (cb_info->callbackSite == CUPTI_API_ENTER) {
      vftr_cuda_region_begin (cbid, cb_info);
   } else if (cb_info->callbackSite == CUPTI_API_EXIT) {
      vftr_cuda_region_end (cbid, cb_info);
   }
}

