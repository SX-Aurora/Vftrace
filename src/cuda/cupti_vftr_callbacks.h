#ifndef CUDA_VFTR_CALLBACKS_H
#define CUDA_VFTR_CALLBACKS_H

// You might think that this file might be better called
// "cupti_callbacks.h" to better suit the names of the other 
// files in this directory. However, there is a file named
// "cupti_callbacks.h" in the cuda include directory, which 
// is also included by the statement below. To avoid this
// collision, we have added the "vftr" interfix.

#include <cupti.h>

void CUPTIAPI vftr_cupti_event_callback (void *userdata,
                                         CUpti_CallbackDomain domain,
                                         CUpti_CallbackId cbid,
                                         const CUpti_CallbackData *cb_info);

enum {T_CUDA_COMP, T_CUDA_MEMCP, T_CUDA_OTHER};
enum {CUDA_NOCOPY=-1, CUDA_COPY_IN=0, CUDA_COPY_OUT=1};

#endif
