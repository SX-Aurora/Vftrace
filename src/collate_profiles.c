#include <stdlib.h>

#ifdef _MPI
#include <mpi.h>
#endif

#include "self_profile.h"
#include "stack_types.h"
#include "collated_stack_types.h"
#include "collate_callprofiles.h"
#include "collate_hwprofiles.h"
#ifdef _MPI
#include "collate_mpiprofiles.h"
#endif
#ifdef _OMP
#include "collate_ompprofiles.h"
#endif
#ifdef _CUDA
#include "collate_cudaprofiles.h"
#endif
#ifdef _ACCPROF
#include "collate_accprofiles.h"
#endif

void vftr_collate_profiles(collated_stacktree_t *collstacktree_ptr,
                           stacktree_t *stacktree_ptr) {
   SELF_PROFILE_START_FUNCTION;
   int myrank = 0;
   int nranks = 1;
   int nstacks = stacktree_ptr->nstacks;
   int *nremote_stacks = NULL;
   myrank = 0;
   nranks = 1;

#ifdef _MPI
   int mpi_initialized;
   PMPI_Initialized(&mpi_initialized);
   if (mpi_initialized) {
      PMPI_Comm_rank(MPI_COMM_WORLD, &myrank);
      PMPI_Comm_size(MPI_COMM_WORLD, &nranks);
   }
#endif

   // get the number of profiles to collect from each rank
   if (myrank == 0) {
      nremote_stacks = (int*) malloc(nranks * sizeof(int));
      nremote_stacks[0] = nstacks;
   }
#ifdef _MPI
   if (mpi_initialized) {
      PMPI_Gather(&nstacks, 1,
                  MPI_INT,
                  nremote_stacks, 1,
                  MPI_INT,
                  0, MPI_COMM_WORLD);
   }
#endif

   vftr_collate_callprofiles(collstacktree_ptr, stacktree_ptr,
                             myrank, nranks, nremote_stacks);
   vftr_collate_hwprofiles (collstacktree_ptr, stacktree_ptr,
                             myrank, nranks, nremote_stacks);
#ifdef _MPI
   vftr_collate_mpiprofiles(collstacktree_ptr, stacktree_ptr,
                            myrank, nranks, nremote_stacks);
#endif

#ifdef _OMP
   vftr_collate_ompprofiles(collstacktree_ptr, stacktree_ptr,
                            myrank, nranks, nremote_stacks);
#endif

#ifdef _CUDA
   vftr_collate_cudaprofiles(collstacktree_ptr, stacktree_ptr,
                              myrank, nranks, nremote_stacks);
#endif
#ifdef _ACCPROF
  vftr_collate_accprofiles (collstacktree_ptr, stacktree_ptr,
                            myrank, nranks, nremote_stacks);
#endif

   free(nremote_stacks);
   SELF_PROFILE_END_FUNCTION;
}
