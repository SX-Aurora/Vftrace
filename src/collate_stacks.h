#ifndef COLLATE_STACKS_H
#define COLLATE_STACKS_H

#include <stdio.h>
#include <stdlib.h>

#include "stack_types.h"
#include "collated_stack_types.h"

collated_stacktree_t vftr_new_empty_collated_stacktree();

collated_stacktree_t vftr_collate_stacks(stacktree_t *stacktree_ptr);

void vftr_collated_stacktree_realloc(collated_stacktree_t *stacktree_ptr);

void vftr_collated_stacktree_free(collated_stacktree_t *stacktree_ptr);

collated_stacktree_t vftr_collated_stacktree_group_by_name(
   collated_stacktree_t *stacktree_ptr);

void vftr_print_collated_stacklist(FILE *fp, collated_stacktree_t stacktree);

char *vftr_get_collated_stack_string(collated_stacktree_t stacktree,
                                     int stackid, bool show_precise);

#endif
