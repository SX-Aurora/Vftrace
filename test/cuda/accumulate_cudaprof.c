#include <stdlib.h>
#include <stdio.h>

#include "stack_types.h"
#include "collated_stack_types.h"
#include "collate_stacks.h"
#include "collate_profiles.h"
#include "callprofiling_types.h"

#include "cupti_vftr_callbacks.h"
#include "cudaprofiling_types.h"

#include "dummy_stacktree.h"

// COMPUTE_CBIDS: CUPTI_RUNTIME_TRACE_CBID_cudaLaunch_v3020,
//                CUPTI_RUNTIME_TRACE_CBID_cudaLaunchKernel_v7000
// MEMCPY_CBIDS:  CUPTI_RUNTIME_TRACE_CBID_cudaMemcpy_v3020,
//                CUPTI_RUNTIME_TRACE_CBID_cudaMemcpyAsync_v3020

int main(int argc, char **argv) {

   vftr_init_dummy_stacktree (10);
   vftr_register_dummy_call_stack ("func0<init", 1);
   vftr_register_dummy_call_stack ("cudafunc1<init", 2);
   vftr_register_dummy_cuda_stack ("cudafunc1<init", CUPTI_RUNTIME_TRACE_CBID_cudaLaunch_v3020,
                                    1000.0, CUDA_NOCOPY, 0);
   vftr_register_dummy_cuda_stack ("cudafunc2<func0<init", CUPTI_RUNTIME_TRACE_CBID_cudaMemcpy_v3020,
				    1000.0, CUDA_COPY_IN, 1048576);
   vftr_register_dummy_cuda_stack ("cudafunc3<func0<init", CUPTI_RUNTIME_TRACE_CBID_cudaMemcpy_v3020,
				    20000.0, CUDA_COPY_OUT, 524288);
   vftr_register_dummy_call_stack ("cudafunc1<init", 2);
   for (int i = 0; i < 4; i++) {
      vftr_register_dummy_cuda_stack ("cudafunc1<init", CUPTI_RUNTIME_TRACE_CBID_cudaLaunch_v3020,
                                       1000.0, CUDA_NOCOPY, 0);
   }
   for (int i = 0; i < 3; i++) {
      vftr_register_dummy_cuda_stack ("cudafunc4<func0<init", CUPTI_RUNTIME_TRACE_CBID_cudaLaunch_v3020,
				       10000.0, CUDA_NOCOPY, 0);
   }



   stacktree_t stacktree = vftr_get_dummy_stacktree();
   
   for (int istack = 0; istack < stacktree.nstacks; istack++) {
      stack_t this_stack = stacktree.stacks[istack];
      profile_t *this_profile = this_stack.profiling.profiles;
      callprofile_t callprof = this_profile->callprof; 
      cudaprofile_t cudaprof = this_profile->cudaprof;
      printf ("istack: %d\n", istack);
      printf ("localID: %d\n", this_stack.lid);
      printf ("name: %s\n", this_stack.name);
      printf ("callprof.t_excl: %lld\n", callprof.time_excl_nsec);
      printf ("callprof.calls: %d\n", callprof.calls);
      printf ("cudaprof.cbid: %d\n", cudaprof.cbid);
      printf ("cudaprof.calls: %d\n", cudaprof.n_calls);
      printf ("cudaprof.t_ms: %.3f\n", cudaprof.t_ms);
      printf ("cudaprof.memcpy_in: %lld\n", cudaprof.memcpy_bytes[CUDA_COPY_IN]);
      printf ("cudaprof.memcpy_out: %lld\n", cudaprof.memcpy_bytes[CUDA_COPY_OUT]);
      printf ("------------------------------------\n");
   }
}
