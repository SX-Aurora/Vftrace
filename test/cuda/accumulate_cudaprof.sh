#!/bin/bash
set -x
test_name=accumulate_cudaprof
output_file=${test_name}.out
ref_file=${srcdir}/ref_output/${test_name}.out

rm -f ${output_file}

./accumulate_cudaprof > ${output_file} || exit 1

diff ${output_file} ${ref_file}
