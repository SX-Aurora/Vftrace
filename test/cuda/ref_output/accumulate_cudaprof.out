istack: 0
localID: 0
name: init
callprof.t_excl: 0
callprof.calls: 1
cudaprof.cbid: 0
cudaprof.calls: 0
cudaprof.t_ms: 0.000
cudaprof.memcpy_in: 0
cudaprof.memcpy_out: 0
------------------------------------
istack: 1
localID: 1
name: func0
callprof.t_excl: 1
callprof.calls: 1
cudaprof.cbid: 0
cudaprof.calls: 0
cudaprof.t_ms: 0.000
cudaprof.memcpy_in: 0
cudaprof.memcpy_out: 0
------------------------------------
istack: 2
localID: 2
name: cudafunc1
callprof.t_excl: 4
callprof.calls: 7
cudaprof.cbid: 13
cudaprof.calls: 5
cudaprof.t_ms: 5000.000
cudaprof.memcpy_in: 0
cudaprof.memcpy_out: 0
------------------------------------
istack: 3
localID: 3
name: cudafunc2
callprof.t_excl: 0
callprof.calls: 1
cudaprof.cbid: 31
cudaprof.calls: 1
cudaprof.t_ms: 1000.000
cudaprof.memcpy_in: 1048576
cudaprof.memcpy_out: 0
------------------------------------
istack: 4
localID: 4
name: cudafunc3
callprof.t_excl: 0
callprof.calls: 1
cudaprof.cbid: 31
cudaprof.calls: 1
cudaprof.t_ms: 20000.000
cudaprof.memcpy_in: 0
cudaprof.memcpy_out: 524288
------------------------------------
istack: 5
localID: 5
name: cudafunc4
callprof.t_excl: 0
callprof.calls: 3
cudaprof.cbid: 13
cudaprof.calls: 3
cudaprof.t_ms: 30000.000
cudaprof.memcpy_in: 0
cudaprof.memcpy_out: 0
------------------------------------
